xds/xds_client_test.go

```diff
@@ -0,0 +1,120 @@
+	"google.golang.org/grpc/xds/clients/xdsclient/xdsresource"
+	"google.golang.org/grpc/xds/clients/xdsclient/xdsresource/version"
+)
+
+type listenerUpdate struct {
+	update any
+	err    error
+}
+
+type listenerWatcher struct {
+	updateCh *testutils.Channel
+}
+
+func newListenerWatcher() *listenerWatcher {
+	return &listenerWatcher{updateCh: testutils.NewChannel()}
+}
+
+func (lw *listenerWatcher) OnUpdate(update xdsclient.ResourceData, onDone xdsclient.OnResourceProcessed) {
+	lw.updateCh.Send(listenerUpdate{update: update.Raw()})
+	onDone()
+}
+
+func (lw *listenerWatcher) OnError(err error, onDone xdsclient.OnResourceProcessed) {
+	// When used with a go-control-plane management server that continuously
+	// resends resources which are NACKed by the xDS client, using a `Replace()`
+	// here and in OnResourceDoesNotExist() simplifies tests which will have
+	// access to the most recently received error.
+	lw.updateCh.Replace(listenerUpdate{err: err})
+	onDone()
+}
+
+func (lw *listenerWatcher) OnResourceDoesNotExist(onDone xdsclient.OnResourceProcessed) {
+	lw.updateCh.Replace(listenerUpdate{err: xdsresource.NewErrorf(xdsresource.ErrorTypeResourceNotFound, "Listener not found in received response")})
+	onDone()
+}
+
+// listenerResourceType provides the resource-type specific functionality for a
+// Listener resource.
+//
+// Implements the xdsclient.ResourceType interface.
+type listenerResourceType struct{}
+
+// Decode deserializes and validates an xDS resource serialized inside the
+// provided `Any` proto, as received from the xDS management server.
+func (listenerResourceType) Decode(opts xdsclient.DecodeOptions, resource any) (*xdsclient.DecodeResult, error) {
+	return nil, nil
+}
+
+func (listenerResourceType) AllResourcesRequiredInSotW() bool {
+	return true
+}
+
+func (listenerResourceType) TypeName() string {
+	return "ListenerResource"
+}
+
+func (listenerResourceType) TypeURL() string {
+	return version.V3ListenerURL
+}
+
+func TestLDSWatch(t *testing.T) {
+	ldsName := "xdsclient-test-lds-resource"
+	rdsName := "xdsclient-test-rds-resource"
+
+	resource := e2e.DefaultClientListener(ldsName, rdsName)
+
+	mgmtServer := e2e.StartManagementServer(t, e2e.ManagementServerOptions{})
+
+	node := clients.Node{ID: uuid.New().String()}
+	defaultAthourity := clients.Authority{XDSServers: []clients.ServerConfig{{ServerURI: mgmtServer.Address}}}
+	authorities := map[string]clients.Authority{"": defaultAthourity}
+
+	serverConfig := clients.ServerConfig{ServerURI: mgmtServer.Address}
+	grpcServerConfig := grpctransport.ServerConfig{Credentials: insecure.NewBundle()}
+	serverConfig.Extensions = grpcServerConfig
+	grpcTransportBuilder := &grpctransport.Builder{}
+
+	listenerResourceType := listenerResourceType{}
+	resourceTypes := map[string]xdsclient.ResourceType{}
+	resourceTypes[listenerResourceType.TypeURL()] = listenerResourceType
+
+	config := xdsclient.NewConfig(defaultAthourity, authorities, node, grpcTransportBuilder, resourceTypes)
```

Comment:
Why is `NewConfig()` taking `defaultAuthority` as a parameter?  There should not be a default authority as a first-class field in the config, only a default list of xDS servers -- I don't think we want to support any extensions of `clients.Authority` at the top level.

Reply:
`Authority` here _is_ apparently just a list of servers:

https://github.com/grpc/grpc-go/pull/7867/files/c81bb55cc5c3110790985aad0a091a5d07a1a3c9#diff-cc980a8a61329b02bc44b7437d28b5fa6741fb5758a1567ba626326d6a18324eR41

Reply:
I think its more clear if we just keep it as a []ServerConfig or a separate struct `ServerConfigs` which holds the slice

Reply:
Also, may be authority doesn't need Extensions since it doesn't get passed around to implementation plugins

Reply:
The authority extensions don't get passed through XdsClient, but they are used outside of it.  The listener resource name template fields are used in both the client and server code that calls XdsClient.

Reply:
Kept the config in DecodeOptions